### PR TITLE
Remove non-BOM components from Altium BOM

### LINKED ISSUE #82: Use Component Properties for Fitted vs not-fitted in BOM Generation

### PR DESCRIPTION
**Pull request by shrik450**
Closes #82. 

Uses the component kind of an altium component to determine if it shouldn't be in the BOM. This is configurable ~and defaults to not removing components to keep the default output the same before and after this change. We can possibly change this to remove by default, but that would be a breaking change.~ This has been changed to remove components by default, see https://github.com/AllSpiceIO/py-allspice/pull/119#issuecomment-2136152414.

<details>
<summary>Outdated</summary>

This is currently a draft as the associated change in Tools has to be deployed to the docker image so I can write a test that won't fail in CI.

What's left:

- [x] Test
</details>

### CODE REVIEW

**Comment by shrik450** [comment]
> This is configurable, and defaults to not removing components to keep the default output the same before and after this change

Expanding on this a little:

Before this PR, if you called generate_bom on a project that had components which were of the No BOM type, we would still include those in the BOM.

With this PR, we have two choices:

1. We remove No BOM components by default, and let users disable the filtering if they want to;
2. We keep No BOM components by default, and let users enable the filtering if they want to.

The advantage of (1) is that this is what most users will expect from BOM generation, as components of No BOM types shouldn't be on the BOM. But the disadvantage is that this is a breaking change: if you ran the same action with the same function in with v3.0 and this PR, you would get different results. 

The advantage of (2) is that it is not a breaking change, but the behavior of `generate_bom` is counter intuitive. 

Another thing is that the generat-bom packaged action could have different defaults. For example, if you use the action in a workflow, it could set the default to True instead of False, and so we don't have to make a breaking change in py-allspice and isolate the breaking change to the action.

@jtran do you have any preferences here?

**Comment by jtran** [comment]
> The advantage of (1) is that this is what most users will expect from BOM generation, as components of No BOM types shouldn't be on the BOM.

If it's expected by the majority of users to remove No BOM components, we should do that.

> But the disadvantage is that this is a breaking change

Arguably, it's a bug fix. But if you want to be stricter about it, why not bump the major version to signal that it's a breaking change?

**Comment by shrik450** [comment]
I've toggled the default. After thinking about it, I agree that it's a bugfix, and the output format removing components that never should've been there shouldn't be "breaking". This will come up again soon with multi-symbol components.

File: allspice/utils/bom_generation.py [hunk 1]
```diff
@@ -455,6 +466,10 @@ def _component_attributes(component: dict) -> ComponentAttributes:
         attributes["_unique_id"] = component["unique_id"]
     except KeyError:
         pass
+    try:
```
Comment (jtran) [inline allspice/utils/bom_generation.py:469]: Should probably update the doc comment for the function that says "This also adds two properties of the component".

File: allspice/utils/bom_generation.py [hunk 2]
```diff
@@ -737,3 +752,15 @@ def _apply_variations(
             final_components.append(component)
 
     return final_components
+
+
+def _remove_non_bom_components(components):
```
Comment (jtran) [inline allspice/utils/bom_generation.py:757]: Type annotations here?

File: tests/test_utils.py [hunk 3]
```diff
@@ -233,6 +235,10 @@ def test_bom_generation_with_default_variant(request, instance):
         "Archimajor.PrjPcb",
         attributes_mapping,
         ref="916e739f3ad9d956f4e2a293542050e1df9e6f9e",
+        # For the variants tests, we don't want to remove non-BOM components
+        # because some of them are enabled by the variants, and we want to
+        # test that they are included when required.
+        remove_non_bom_components=False,
```
Comment (jtran) [inline tests/test_utils.py:241]: Would users using variants need to think about this? Or is it just us for testing?

Comment (shrik450) [reply tests/test_utils.py:241]: Just for us testing, so you can visually confirm that the correct number of components are removed when you compare with the project in Altium Designer or open the PrjPcb file.

File: allspice/utils/bom_generation.py (hunk 2 above)
Comment (shrik450) [reply allspice/utils/bom_generation.py]: Added.

File: allspice/utils/bom_generation.py (hunk 1 above)
Comment (jtran) [reply allspice/utils/bom_generation.py:481]: I meant at the top of this function. But as long as it's on your radar, that's fine.

**Review by jtran — APPROVED** [review]

**Review by kdumontnu — APPROVED** [review]
🙌

**Pull request merged**